plugins/node/opentelemetry-hapi-instrumentation/src/hapi.ts

```diff
@@ -0,0 +1,373 @@
+export class HapiInstrumentation extends BasePlugin<typeof Hapi> {
+  static readonly component = HapiComponentName;
+
+  constructor(readonly moduleName: string) {
+    super('@opentelemetry/hapi-instrumentation', VERSION);
+  }
+
+  /**
+   * Patches Hapi operations by wrapping the Hapi.server and Hapi.Server functions
+   */
+  protected patch(): typeof Hapi {
+    this._logger.debug('Patching Hapi');
+    if (this._moduleExports == null) {
+      return this._moduleExports;
+    }
+
+    this._logger.debug('Patching Hapi.server');
+    shimmer.wrap(
+      this._moduleExports,
+      'server',
+      this._getServerPatch.bind(this)
+    );
+
+    // Casting as any is necessary here due to an issue with the @types/hapi__hapi
+    // type definition for Hapi.Server. Hapi.Server (note the uppercase) can also function
+    // as a factory function, similarly to Hapi.server (lowercase), and so should
+    // also be supported and instrumented. This is an issue with the DefinitelyTyped repo.
+    // Function is defined at: https://github.com/hapijs/hapi/blob/master/lib/index.js#L9
+    this._logger.debug('Patching Hapi.Server');
+    shimmer.wrap(
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      this._moduleExports as any,
+      'Server',
+      this._getServerPatch.bind(this)
+    );
+
+    return this._moduleExports;
+  }
+
+  /**
+   * Unpatches all Hapi operations
+   */
+  protected unpatch(): void {
+    this._logger.debug('Unpatching Hapi');
+    shimmer.massUnwrap([this._moduleExports], ['server', 'Server']);
+  }
+
+  /**
+   * Patches the Hapi.server and Hapi.Server functions in order to instrument
+   * the server.route, server.ext, and server.register functions via calls to the
+   * @function _getServerRoutePatch, @function _getServerExtPatch, and
+   * @function _getServerRegisterPatch functions
+   * @param original - the original Hapi Server creation function
+   */
+  private _getServerPatch(
+    original: (options?: Hapi.ServerOptions) => Hapi.Server
+  ) {
+    const instrumentation = this;
+    return function server(this: Hapi.Server, opts?: Hapi.ServerOptions) {
+      const newServer: Hapi.Server = original.apply(this, [opts]);
+
+      shimmer.wrap(newServer, 'route', original => {
+        return instrumentation._getServerRoutePatch.bind(instrumentation)(
+          original
+        );
+      });
+
+      shimmer.wrap(newServer, 'ext', original => {
+        return instrumentation._getServerExtPatch.bind(instrumentation)(
+          original
+        );
+      });
+
+      // Casting as any is necessary here due to multiple overloads on the Hapi.Server.register
+      // function, which requires supporting a variety of different types of Plugin inputs
+      shimmer.wrap(
+        newServer,
+        'register',
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        instrumentation._getServerRegisterPatch.bind(instrumentation) as any
```

Comment:
Do you not need to pass in original here?

Reply:
Here it's passed in automatically by the shimmer function!